File: 이것이 코딩테스트다/03.DFS&BFS/6.음료수 얼려 먹기.py

n , m = map(int, input().split())
graph = []
for i in range(n):
  graph.append(list(map(int, input().split())))
result = 0
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]


def dfs(x, y):
  graph[x][y] = 1

  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]

    if (nx >= 0 and ny >= 0 and nx < n and ny <m):
      if graph[nx][ny] == 0:
        dfs(nx, ny)

for i in range(n):
  for j in range(m):
    if graph[i][j] == 0:
      dfs(i, j)
      result += 1

print(result)

File: 이것이 코딩테스트다/03.DFS&BFS/6.음료수 얼려 먹기.js
// 조합 => DFS
const text = ["00110", "00011", "11111", "00000"];

const graph = text.map((t) => t.split("").map(Number));
let result = 0;
const n = graph.length;
const m = graph[0].length;
const dx = [-1, 0, 1, 0];
const dy = [0, 1, 0, -1];

const dfs = (x, y) => {
  graph[x][y] = 1;

  for (let i = 0; i < 4; i++) {
    const nx = x + dx[i];
    const ny = y + dy[i];

    if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
      if (graph[nx][ny] === 0) {
        dfs(nx, ny);
      }
    }
  }
};

for (let i = 0; i < n; i++) {
  for (let j = 0; j < m; j++) {
    if (graph[i][j] === 0) {
      dfs(i, j);
      result += 1;
    }
  }
}
console.log(result);
